provider/pkg/schema/gen_tags.go

```diff
@@ -0,0 +1,35 @@
+package schema
+
+import (
+	jsschema "github.com/lestrrat-go/jsschema"
+	pschema "github.com/pulumi/pulumi/pkg/v3/codegen/schema"
+)
+
+type TagsStyle string
+
+const (
+	TagsStyleUnknown       TagsStyle = ""
+	TagsStyleUntyped       TagsStyle = "untyped"
+	TagsStyleStringMap     TagsStyle = "stringMap"
+	TagsStyleKeyValueArray TagsStyle = "keyValueArray"
+)
+
+func (ctx *context) GetTagsStyle(typeSpec *pschema.TypeSpec, originalSpec *jsschema.Schema) TagsStyle {
```

Comment:
Maybe complex enough to warrant a table-based test, especially if there will be more cases later.